File: src/jira_api/services/issue_service.py
"""Issue service for JIRA operations."""

from typing import List, Optional

from jira_api.core.client import JiraClient
from jira_api.models.issue import (
    Issue,
    IssueAssignment,
    IssueCreate,
    IssueTransition,
    IssueTransitionRequest,
    IssueUpdate,
)
from jira_api.services.user_service import UserService


class IssueService:
    """Service for issue-related operations."""

    def __init__(self, client: JiraClient) -> None:
        """Initialize the issue service.

        Args:
            client: JIRA API client instance
        """
        self.client = client
        self.user_service = UserService(client)

    def create_issue(
        self,
        project_id: str,
        summary: str,
        issue_type_id: str,
        description: Optional[str] = None,
        priority_id: Optional[str] = None,
        assignee_email: Optional[str] = None,
        labels: Optional[List[str]] = None,
    ) -> Issue:
        """Create a new issue with simplified parameters.

        Args:
            project_id: ID of the project
            summary: Summary of the issue
            issue_type_id: ID of the issue type
            description: Optional description
            priority_id: Optional priority ID
            assignee_email: Optional email of the assignee
            labels: Optional list of labels

        Returns:
            Created Issue object
        """
        assignee_account_id = None
        if assignee_email:
            assignee = self.user_service.get_user_by_email(assignee_email)
            if assignee:
                assignee_account_id = assignee.account_id

        issue_data = IssueCreate(
            project_id=project_id,
            summary=summary,
            description=description,
            issue_type_id=issue_type_id,
            priority_id=priority_id,
            assignee_account_id=assignee_account_id,
            labels=labels or [],
        )

        return self.client.create_issue(issue_data)

    def create_issue_by_type_name(
        self,
        project_key: str,
        summary: str,
        issue_type_name: str,
        description: Optional[str] = None,
        priority_id: Optional[str] = None,
        assignee_email: Optional[str] = None,
        labels: Optional[List[str]] = None,
    ) -> Issue:
        """Create a new issue using issue type name instead of ID.

        Args:
            project_key: Key of the project
            summary: Summary of the issue
            issue_type_name: Name of the issue type (e.g., 'Bug', 'Story', 'Task')
            description: Optional description
            priority_id: Optional priority ID
            assignee_email: Optional email of the assignee
            labels: Optional list of labels

        Returns:
            Created Issue object

        Raises:
            JiraValidationError: If issue type name is not found in project
        """
        # Get the project to find the project ID
        project = self.client.get_project(project_key)
        project_id = project.id
        
        # Get issue type ID by name
        issue_type_id = self.client.get_issue_type_id_by_name(project_key, issue_type_name)
        
        return self.create_issue(
            project_id=project_id,
            summary=summary,
            issue_type_id=issue_type_id,
            description=description,
            priority_id=priority_id,
            assignee_email=assignee_email,
            labels=labels,
        )

    def get_issue(self, issue_key: str) -> Issue:
        """Get an issue by its key.

        Args:
            issue_key: The issue key (e.g., PROJ-123)

        Returns:
            Issue object
        """
        return self.client.get_issue(issue_key)

    def update_issue_summary(self, issue_key: str, summary: str) -> None:
        """Update an issue's summary.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            summary: New summary text
        """
        update_data = IssueUpdate(summary=summary)
        self.client.update_issue(issue_key, update_data)

    def update_issue_description(self, issue_key: str, description: str) -> None:
        """Update an issue's description.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            description: New description text
        """
        update_data = IssueUpdate(description=description)
        self.client.update_issue(issue_key, update_data)

    def add_labels_to_issue(self, issue_key: str, labels: List[str]) -> None:
        """Add labels to an issue.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            labels: List of labels to add
        """
        update_data = IssueUpdate(labels_add=labels)
        self.client.update_issue(issue_key, update_data)

    def remove_labels_from_issue(self, issue_key: str, labels: List[str]) -> None:
        """Remove labels from an issue.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            labels: List of labels to remove
        """
        update_data = IssueUpdate(labels_remove=labels)
        self.client.update_issue(issue_key, update_data)

    def assign_issue_by_email(self, issue_key: str, assignee_email: str) -> None:
        """Assign an issue to a user by email.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            assignee_email: Email of the user to assign the issue to
        """
        assignee = self.user_service.get_user_by_email(assignee_email)
        if not assignee:
            raise ValueError(f"User with email '{assignee_email}' not found")

        assignment = IssueAssignment(account_id=assignee.account_id)
        self.client.assign_issue(issue_key, assignment)

    def unassign_issue(self, issue_key: str) -> None:
        """Unassign an issue.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
        """
        assignment = IssueAssignment(account_id=None)
        self.client.assign_issue(issue_key, assignment)

    def get_available_transitions(self, issue_key: str) -> List[IssueTransition]:
        """Get available transitions for an issue.

        Args:
            issue_key: The issue key (e.g., PROJ-123)

        Returns:
            List of available transitions
        """
        return self.client.get_issue_transitions(issue_key)

    def transition_issue_by_name(
        self,
        issue_key: str,
        transition_name: str,
        comment: Optional[str] = None,
        resolution_name: Optional[str] = None,
    ) -> None:
        """Transition an issue by transition name.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            transition_name: Name of the transition to perform
            comment: Optional comment to add with the transition
            resolution_name: Optional resolution name

        Raises:
            ValueError: If transition name is not found
        """
        transitions = self.get_available_transitions(issue_key)
        
        transition_id = None
        for transition in transitions:
            if transition.name.lower() == transition_name.lower():
                transition_id = transition.id
                break

        if not transition_id:
            available_names = [t.name for t in transitions]
            raise ValueError(
                f"Transition '{transition_name}' not found. "
                f"Available transitions: {', '.join(available_names)}"
            )

        transition_request = IssueTransitionRequest(
            transition_id=transition_id,
            comment=comment,
            resolution_name=resolution_name,
        )
        
        self.client.transition_issue(issue_key, transition_request)

    def transition_issue_by_id(
        self,
        issue_key: str,
        transition_id: str,
        comment: Optional[str] = None,
        resolution_name: Optional[str] = None,
    ) -> None:
        """Transition an issue by transition ID.

        Args:
            issue_key: The issue key (e.g., PROJ-123)
            transition_id: ID of the transition to perform
            comment: Optional comment to add with the transition
            resolution_name: Optional resolution name
        """
        transition_request = IssueTransitionRequest(
            transition_id=transition_id,
            comment=comment,
            resolution_name=resolution_name,
        )
        
        self.client.transition_issue(issue_key, transition_request)

File: node/src/services/issue.service.ts
/**
 * IssueService - High-level issue operations
 * Provides human-readable type names, email-based assignment, and bulk operations
 */

import type { JiraClient } from '../core/client.js';
import type { UserService } from './user.service.js';
import type {
  Issue,
  IssueCreateInput,
  IssueCreate,
  IssueUpdateInput,
  IssueUpdate,
  IssueTransition,
  IssueFields,
} from '../models/issue.js';
import { JiraNotFoundError, JiraValidationError } from '../errors/index.js';
import { getLogger } from '../utils/logger.js';
import { textToADF } from '../utils/adf.js';

/**
 * Bulk create result
 */
export interface BulkCreateResult {
  successful: Issue[];
  failed: Array<{ input: IssueCreateInput; error: Error }>;
}

/**
 * Bulk update result
 */
export interface BulkUpdateResult {
  successful: string[];
  failed: Array<{ key: string; error: Error }>;
}

/**
 * IssueService class
 * Provides business logic layer for issue operations
 */
export class IssueService {
  private logger = getLogger();

  constructor(
    private client: JiraClient,
    private userService: UserService
  ) {}

  // ======================
  // CREATION METHODS
  // ======================

  /**
   * Create issue with human-readable type name
   * Automatically resolves issue type name to ID
   */
  async createIssueByTypeName(data: IssueCreateInput): Promise<Issue> {
    this.logger.debug({ data }, 'Creating issue by type name');

    // Validate required fields
    if (!data.projectKey || !data.issueTypeName || !data.summary) {
      throw new JiraValidationError(
        'projectKey, issueTypeName, and summary are required'
      );
    }

    // Resolve issue type name to ID
    const issueTypeId = await this.client.getIssueTypeIdByName(
      data.projectKey,
      data.issueTypeName
    );

    // Resolve assignee email to account ID if provided
    let assigneeId: string | undefined;
    if (data.assigneeEmail) {
      assigneeId = await this.userService.resolveToAccountId(data.assigneeEmail);
    }

    // Build API request
    const createRequest: IssueCreate = {
      fields: {
        project: { key: data.projectKey },
        issuetype: { id: issueTypeId },
        summary: data.summary,
        description: data.description ? textToADF(data.description) : undefined,
        assignee: assigneeId ? { accountId: assigneeId } : undefined,
        priority: data.priority ? { name: data.priority } : undefined,
        labels: data.labels || [],
        parent: data.parentKey ? { key: data.parentKey } : undefined,
      },
    };

    return this.client.createIssue(createRequest);
  }

  /**
   * Create issue with type ID (low-level)
   */
  async createIssue(data: IssueCreate): Promise<Issue> {
    this.logger.debug('Creating issue with type ID');
    return this.client.createIssue(data);
  }

  /**
   * Create bug (convenience method)
   */
  async createBug(
    projectKey: string,
    summary: string,
    description?: string,
    assigneeEmail?: string
  ): Promise<Issue> {
    return this.createIssueByTypeName({
      projectKey,
      issueTypeName: 'Bug',
      summary,
      description,
      assigneeEmail,
    });
  }

  /**
   * Create task (convenience method)
   */
  async createTask(
    projectKey: string,
    summary: string,
    description?: string,
    assigneeEmail?: string
  ): Promise<Issue> {
    return this.createIssueByTypeName({
      projectKey,
      issueTypeName: 'Task',
      summary,
      description,
      assigneeEmail,
    });
  }

  /**
   * Create story (convenience method)
   */
  async createStory(
    projectKey: string,
    summary: string,
    description?: string,
    assigneeEmail?: string
  ): Promise<Issue> {
    return this.createIssueByTypeName({
      projectKey,
      issueTypeName: 'Story',
      summary,
      description,
      assigneeEmail,
    });
  }

  // ======================
  // RETRIEVAL METHODS
  // ======================

  /**
   * Get issue by key
   */
  async getIssue(key: string, fields?: string[]): Promise<Issue> {
    this.logger.debug({ key, fields }, 'Getting issue');
    return this.client.getIssue(key, fields);
  }

  /**
   * Search issues with JQL
   */
  async searchIssues(jql: string, maxResults: number = 50): Promise<Issue[]> {
    this.logger.debug({ jql, maxResults }, 'Searching issues');
    const result = await this.client.searchIssues(jql, maxResults);
    return result.issues || [];
  }

  // ======================
  // UPDATE METHODS
  // ======================

  /**
   * Update issue summary
   */
  async updateIssueSummary(key: string, summary: string): Promise<void> {
    this.logger.debug({ key, summary }, 'Updating issue summary');
    await this.client.updateIssue(key, {
      fields: { summary },
    });
  }

  /**
   * Update issue description
   */
  async updateIssueDescription(key: string, description: string): Promise<void> {
    this.logger.debug({ key }, 'Updating issue description');
    await this.client.updateIssue(key, {
      fields: { description: textToADF(description) },
    });
  }

  /**
   * Update issue fields
   */
  async updateIssue(key: string, fields: Partial<IssueFields>): Promise<void> {
    this.logger.debug({ key, fields }, 'Updating issue');
    await this.client.updateIssue(key, { fields });
  }

  /**
   * Update issue with input format
   */
  async updateIssueWithInput(key: string, input: IssueUpdateInput): Promise<void> {
    this.logger.debug({ key, input }, 'Updating issue with input');

    const update: IssueUpdate = {
      fields: {},
      update: {},
    };

    // Handle field updates
    if (input.summary) {
      update.fields!.summary = input.summary;
    }

    if (input.description) {
      update.fields!.description = textToADF(input.description);
    }

    // Handle label additions
    if (input.labelsAdd && input.labelsAdd.length > 0) {
      update.update!.labels = input.labelsAdd.map((label) => ({ add: label }));
    }

    // Handle label removals
    if (input.labelsRemove && input.labelsRemove.length > 0) {
      update.update!.labels = update.update!.labels || [];
      update.update!.labels.push(...input.labelsRemove.map((label) => ({ remove: label })));
    }

    await this.client.updateIssue(key, update);
  }

  // ======================
  // LABEL MANAGEMENT
  // ======================

  /**
   * Add labels to issue
   */
  async addLabelsToIssue(key: string, labels: string[]): Promise<void> {
    this.logger.debug({ key, labels }, 'Adding labels to issue');
    await this.client.updateIssue(key, {
      update: {
        labels: labels.map((label) => ({ add: label })),
      },
    });
  }

  /**
   * Remove labels from issue
   */
  async removeLabelsFromIssue(key: string, labels: string[]): Promise<void> {
    this.logger.debug({ key, labels }, 'Removing labels from issue');
    await this.client.updateIssue(key, {
      update: {
        labels: labels.map((label) => ({ remove: label })),
      },
    });
  }

  // ======================
  // ASSIGNMENT METHODS
  // ======================

  /**
   * Assign issue by email
   */
  async assignIssueByEmail(key: string, email: string): Promise<void> {
    this.logger.debug({ key, email }, 'Assigning issue by email');
    const accountId = await this.userService.resolveToAccountId(email);
    await this.client.assignIssue(key, accountId);
  }

  /**
   * Assign issue by account ID
   */
  async assignIssueById(key: string, accountId: string): Promise<void> {
    this.logger.debug({ key, accountId }, 'Assigning issue by ID');
    await this.client.assignIssue(key, accountId);
  }

  /**
   * Unassign issue
   */
  async unassignIssue(key: string): Promise<void> {
    this.logger.debug({ key }, 'Unassigning issue');
    await this.client.assignIssue(key, null);
  }

  // ======================
  // TRANSITION METHODS
  // ======================

  /**
   * Get available transitions for issue
   */
  async getAvailableTransitions(key: string): Promise<IssueTransition[]> {
    this.logger.debug({ key }, 'Getting available transitions');
    return this.client.getIssueTransitions(key);
  }

  /**
   * Transition issue by name
   */
  async transitionIssueByName(
    key: string,
    transitionName: string,
    comment?: string
  ): Promise<void> {
    this.logger.debug({ key, transitionName }, 'Transitioning issue by name');

    // Get available transitions
    const transitions = await this.client.getIssueTransitions(key);

    // Find transition by name (case-insensitive)
    const transition = transitions.find(
      (t) => t.name.toLowerCase() === transitionName.toLowerCase()
    );

    if (!transition) {
      const availableNames = transitions.map((t) => t.name).join(', ');
      throw new JiraNotFoundError(
        `Transition '${transitionName}' not found for issue ${key}. Available: ${availableNames}`
      );
    }

    // Execute transition
    await this.client.transitionIssue(key, {
      transition: { id: transition.id },
      update: comment
        ? {
            comment: [{ add: { body: textToADF(comment) } }],
          }
        : undefined,
    });
  }

  /**
   * Transition issue by ID
   */
  async transitionIssueById(
    key: string,
    transitionId: string,
    comment?: string
  ): Promise<void> {
    this.logger.debug({ key, transitionId }, 'Transitioning issue by ID');

    await this.client.transitionIssue(key, {
      transition: { id: transitionId },
      update: comment
        ? {
            comment: [{ add: { body: textToADF(comment) } }],
          }
        : undefined,
    });
  }

  // ======================
  // BULK OPERATIONS
  // ======================

  /**
   * Bulk create issues
   * Creates multiple issues with concurrency control
   */
  async bulkCreateIssues(
    issues: IssueCreateInput[],
    concurrency: number = 5
  ): Promise<BulkCreateResult> {
    this.logger.info({ count: issues.length, concurrency }, 'Bulk creating issues');

    const successful: Issue[] = [];
    const failed: Array<{ input: IssueCreateInput; error: Error }> = [];

    // Process in batches
    for (let i = 0; i < issues.length; i += concurrency) {
      const batch = issues.slice(i, i + concurrency);
      const results = await Promise.allSettled(
        batch.map((issue) => this.createIssueByTypeName(issue))
      );

      results.forEach((result, index) => {
        if (result.status === 'fulfilled') {
          successful.push(result.value);
        } else {
          failed.push({
            input: batch[index]!,
            error: result.reason as Error,
          });
        }
      });

      this.logger.debug({ processed: i + batch.length, total: issues.length }, 'Batch complete');
    }

    this.logger.info(
      { successful: successful.length, failed: failed.length },
      'Bulk create complete'
    );

    return { successful, failed };
  }

  /**
   * Bulk update issues
   */
  async bulkUpdateIssues(
    updates: Array<{ key: string; fields: Partial<IssueFields> }>,
    concurrency: number = 5
  ): Promise<BulkUpdateResult> {
    this.logger.info({ count: updates.length, concurrency }, 'Bulk updating issues');

    const successful: string[] = [];
    const failed: Array<{ key: string; error: Error }> = [];

    // Process in batches
    for (let i = 0; i < updates.length; i += concurrency) {
      const batch = updates.slice(i, i + concurrency);
      const results = await Promise.allSettled(
        batch.map((update) => this.updateIssue(update.key, update.fields))
      );

      results.forEach((result, index) => {
        if (result.status === 'fulfilled') {
          successful.push(batch[index]!.key);
        } else {
          failed.push({
            key: batch[index]!.key,
            error: result.reason as Error,
          });
        }
      });

      this.logger.debug({ processed: i + batch.length, total: updates.length }, 'Batch complete');
    }

    this.logger.info(
      { successful: successful.length, failed: failed.length },
      'Bulk update complete'
    );

    return { successful, failed };
  }
}
